data/api/client-server/account-data.yaml

```diff
@@ -71,6 +71,38 @@ paths:
             application/json: {}
           schema:
             type: object
+        400:
+          description: |-
+            The request body is not a JSON object. Errcode: `M_BAD_JSON`.
```

Comment:
hum. https://spec.matrix.org/v1.3/client-server-api/#common-error-codes says we should use `M_NOT_JSON` in this case. If synapse is using `M_BAD_JSON`, I'd say that's a synapse bug. (And yes, those are terrible names for the error codes. Don't blame me.)

(similarly for `PUT /user/{userId}/rooms/{roomId}/account_data/{type}`)

Reply:
Judging from https://github.com/matrix-org/synapse/blob/33abbc327813e65aaa91e10f98a31622c045004c/synapse/http/servlet.py#L671-L694 it looks like Synapse is being naughty.

Reply:
Actually, now I'm not sure if this change is correct. Synapse hits that code when it's parsed a JSON value that isn't an object (so list, bool, null, number or string). They're arguably all "valid JSON", just not the form of JSON that we were expecting?

Reply:
oh right, yes, I'd expect `M_BAD_JSON` for a list where we expect an object. Which means that this endpoint can return either `M_BAD_JSON` or `M_NOT_JSON`.

Reply:
See [b66b199](https://github.com/matrix-org/matrix-spec/pull/1155/commits/b66b1994a0641d7fe8b9cf5978b0a411f580e694)